contracts/NonfungiblePositionManagerV4.sol

```diff
@@ -133,7 +174,47 @@ contract NonfungiblePositionManagerV4 is
         checkDeadline(params.deadline)
         returns (uint256 tokenId, uint128 liquidity, uint256 amount0, uint256 amount1)
     {
-        // TODO: implement this
+        (tokenId, liquidity, amount0, amount1) = abi.decode(
+            poolManager.lock(abi.encode(CallbackData(CallbackAction.Mint, msg.sender, abi.encode(params)))),
+            (uint256, uint128, uint256, uint256)
+        );
+    }
+
+    function mintCallback(MintParams memory params)
+        internal
+        returns (uint256 tokenId, uint128 liquidity, BalanceDelta delta)
+    {
+        PoolId poolId = params.poolKey.toId();
+        (liquidity, delta) = addLiquidity(
+            AddLiquidityParams({
+                poolKey: params.poolKey,
+                tickLower: params.tickLower,
+                tickUpper: params.tickUpper,
+                amount0Desired: params.amount0Desired,
+                amount1Desired: params.amount1Desired,
+                amount0Min: params.amount0Min,
+                amount1Min: params.amount1Min,
+                hookData: params.hookData
+            })
+        );
+
+        tokenId = _nextId++;
+        _mint(params.recipient, tokenId);
+
+        Position.Info memory positionInfo =
+            poolManager.getPosition(poolId, address(this), params.tickLower, params.tickUpper);
+        _positions[tokenId] = TokenPosition({
+            nonce: 0,
```

Comment:
🤔 copied from v3 but not sure why this, operator, tokensOwed are all initialized to 0